core/trino-main/src/main/java/io/trino/operator/output/PositionsAppender.java

```diff
@@ -11,44 +11,25 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-
 package io.trino.operator.output;
 
 import io.trino.spi.block.Block;
-import io.trino.spi.block.BlockBuilder;
-import io.trino.spi.type.Type;
+import io.trino.spi.block.BlockBuilderStatus;
+import io.trino.spi.block.RunLengthEncodedBlock;
 import it.unimi.dsi.fastutil.ints.IntArrayList;
 
-import static java.util.Objects.requireNonNull;
-
 public interface PositionsAppender
```

Comment:
Let's leave `PositionsAppender` as it is (it's purpose is to add selected positions one by one)

Let's introduce another interface, e.g. `BatchAppender` which would be similar to:
```
interface BatchAppender {
  void appendBlock(IntArrayList positions, Block block);
  Block build(); 
}
```
`BatchAppender` would keep state internally (which can include `BlockBuilder`).

You can then have
```
interface BatchAppenderFactory {
  BatchAppender create(Type type);
}
```
which you can use to recursively create batch appenders for primitive and nested types (e.g. `RowType`)

Reply:
In fact, I wonder if we could ditch existing `PositionsAppender` and just use `BatchAppender`. If we handled RLE and dictionary by some `DictionaryAwareBatchAppender`, then we wouldn't need to isolate based on input `Block` type. Hence, `BlockBuilder` argument in `PositionsAppender` becomes irrelevant